### PR TITLE
Use error-chain for root Error type

### PR DESCRIPTION
**Pull request by KodrAus**
Opening this up for visibility, but I've still got plenty of work to do on it before it's ready for review. The plan is:

- Use `error-chain` for the main `tantivy::Error` type
- Ensure all errors implement `std::Error`

### CODE REVIEW

**Comment by fulmicoton** [comment]
Thank you

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,5 @@ target/release
 Cargo.lock
 benchmark
 .DS_Store
-cpp/simdcomp/bitpackingbenchmark
\ No newline at end of file
+cpp/simdcomp/bitpackingbenchmark
+*.bk
```
Comment (fulmicoton) [inline .gitignore:9]: hehe :)

File: Cargo.toml [hunk 2]
```diff
@@ -39,6 +39,7 @@ version = "2"
 crossbeam = "0.2"
 futures = "0.1.9"
 futures-cpupool = "0.1.2"
+error-chain = "0.8.1"
```
Comment (fulmicoton) [inline Cargo.toml]: I'm afraid I'm guilty for that bad habit, but I think we should probably avoid putting minor in there. So 0.8 should be fine. I should have done it for futures, and futures-cpupool as well

File: src/core/index.rs [hunk 3]
```diff
@@ -1,5 +1,5 @@
 use Result;
-use Error;
+use error::*;
```
Comment (fulmicoton) [inline src/core/index.rs]: do we need the * there ? It is only for ErrorKind isn't it?

File: src/error.rs [hunk 4]
```diff
@@ -3,105 +3,129 @@
...
-    ErrorInThread(String),
-    /// An Error appeared related to the lack of a field.
-    SchemaError(String),
-    /// Tried to access a fastfield reader for a field not configured accordingly.
-    FastFieldError(FastFieldNotAvailableError),
-}
+error_chain!(
+    errors {
+        PathDoesNotExist(buf: PathBuf) {
+            description("path does not exist")
```
Comment (fulmicoton) [inline src/error.rs]: have you checked that the rustdoc looked ok? On the first PR, the doc was not generated for the error kind

File: src/indexer/index_writer.rs [hunk 5]
```diff
@@ -9,7 +9,7 @@ use core::SegmentReader;
 use indexer::stamper::Stamper;
 use datastruct::stacker::Heap;
 use directory::FileProtection;
-use Error;
+use error::*;
```
Comment (fulmicoton) [inline src/indexer/index_writer.rs]: no star, if possible

File: src/directory/managed_directory.rs [hunk 6]
```diff
@@ -12,8 +12,7 @@ use std::io::Write;
 use core::MANAGED_FILEPATH;
 use std::collections::HashMap;
 use std::fmt;
-use Result;
-use Error;
+use error::*;
```
Comment (fulmicoton) [inline src/directory/managed_directory.rs]: no star, if possible

File: src/indexer/index_writer.rs [hunk 7]
```diff
@@ -603,7 +601,7 @@ mod tests {
     use schema::{self, Document};
     use Index;
     use Term;
-    use Error;
+    use error::*;
```
Comment (fulmicoton) [inline src/indexer/index_writer.rs:604]: no star, if possible

File: src/indexer/segment_updater.rs [hunk 8]
```diff
@@ -7,7 +7,7 @@ use core::SegmentMeta;
 use core::SerializableSegment;
 use directory::Directory;
 use indexer::stamper::Stamper;
-use Error;
+use error::*;
```
Comment (fulmicoton) [inline src/indexer/segment_updater.rs]: no star, if possible

**Comment by fulmicoton** [comment]
LGTM, added a bunch of minor comments

File: src/core/index.rs (hunk 3 above)
Comment (KodrAus) [reply src/core/index.rs]: There's also a `ResultExt` trait that gives us the `chain_err` method. 

It's not uncommon to `use error::*` but if you'd rather avoid it I think that's reasonable too, I can just import the bits from `error` I actually need.

File: src/directory/error.rs [hunk 9]
```diff
@@ -45,8 +175,46 @@ pub enum DeleteError {
...
+        match *self {
+            DeleteError::FileDoesNotExist(ref path) => {
+                write!(f, "the file '{:?}' does not exist", path)
+            }
+            DeleteError::FileProtected(ref path) => {
+                write!(f, "the file '{:?}' is protected and can't be deleted", path)
+            }
+            DeleteError::IOError(ref err) => {
+                write!(f,
+                       "an io error occurred while opening a file for reading: '{}'",
```
Comment (KodrAus) [inline src/directory/error.rs]: Wrong message

File: src/core/index.rs (hunk 3 above)
Comment (fulmicoton) [reply src/core/index.rs]: I'd rather avoid it.

**Pull request merged**